### DISCUSSION TITLE
How to model interactions with a UIViewRepresentable in TCA?

### DISCUSSION THREAD

**Discussion by SilverTab** [pointfreeco/swift-composable-architecture#2008] [Q&A] [answered]
As the title suggest, I'm looking for suggestions as to how to model interactions with a UIViewRepresentable and its Coordinator so that it gracefully integrates with TCA. I've seen some examples (one in composable-core-location) using bindings but it doesn't quite fit what I'm trying to do here.

For some context; I have a `WKWebView` wrapped in a UIViewRepresentable, and I'm using the webview's javascript bridge to listen to events coming from the web page. The events are handled in a method of the Coordinator. I would like to somehow "plug" these events into a TCA reducer, ideally they would trigger an Action on a Reducer, but I'm not quite sure on the best way to proceed. 

If it's of any help, here's what my WebView and its Coordinator look like:

```swift
struct WebView: UIViewRepresentable {
    class Coordinator: NSObject, WKNavigationDelegate, WKScriptMessageHandler {
        var webView: WKWebView?
        
        func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
            self.webView = webView
        }

        func makeUIView(context: Context) -> WKWebView {
            let coordinator = makeCoordinator()
            let userContentController = WKUserContentController()
            userContentController.add(coordinator, name: "bridge")
            
            let configuration = WKWebViewConfiguration()
            configuration.userContentController = userContentController
            
            let _wkwebview = WKWebView(frame: .zero, configuration: configuration)
            _wkwebview.navigationDelegate = coordinator
            
            return _wkwebview
        }
        
        // receive message from wkwebview
        func userContentController(
            _ userContentController: WKUserContentController,
            didReceive message: WKScriptMessage
        ) {
            // We got a message from the webview!
            // How should I send an action to a TCA store from here?
            print(message.body)
        }
    }
}
```

I've toyed with the ideal of passing a TCA Store to the WebView, but that still doesn't really give me a ViewStore accessible from within the coordinator in order to send actions to the store 🤔 

Any ideas/suggestions would be appreciated!

**Comment by saroar** [comment, marked as answer]
This is how I did 
https://github.com/AddaMeSPB/AddaMeIOS/blob/siteRouter/AddameSPM/Sources/SettingsFeature/TermsAndPrivacyWebView.swift#L21

**Reply by SilverTab** [reply]
Nice! Thanks a lot for the concrete example, this is super useful!

**Reply by saroar** [reply]
its my open source app feel free to use all code :)

**Comment by pyrtsa** [comment]
If you pass `store` to the coordinator, you can send actions to it by constructing a temporary `ViewStore`:
```swift
ViewStore(store.stateless).send(.didReceive(message))
```
PS. If you hold a reference from `Coordinator` to the `WKWebView` instance, remember to either make it `weak` or else break the cycle in your `UIViewRepresentable`'s `static func dismantleUIView(_:coordinator:)` implementation!

**Reply by SilverTab** [reply]
Good call-out! Will use `dismantleUIView` to do the proper cleanup 👍
